pkg/util/watch/until.go

```diff
@@ -0,0 +1,403 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to merge list options")
+	}
+	optsList := UnrollListOptions(cOpts)
+	klog.V(5).Infof("Listing %T %s: %v", objList, objList.GetObjectKind().GroupVersionKind().Kind, optsList)
+	err = fw.Client.List(fw.Context, objList, optsList...)
+	return objList, err
+}
+
+// Watch satisfies the cache.Watcher interface
+func (fw *flexWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	objList, err := kinds.ObjectAsClientObjectList(fw.ObjList.DeepCopyObject())
+	if err != nil {
+		return nil, err
+	}
+	cOpts, err := ConvertListOptions(&options)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert metav1.ListOptions to client.ListOptions")
+	}
+	cOpts, err = MergeListOptions(cOpts, fw.listOptions())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to merge list options")
+	}
+	optsList := UnrollListOptions(cOpts)
+	klog.V(5).Infof("Watching %T %s: %v", objList, objList.GetObjectKind().GroupVersionKind().Kind, optsList)
+	return fw.Client.Watch(fw.Context, objList, optsList...)
+}
+
+// listOptions builds a set of List filters for the flexWatcher's Key and Labels.
+func (fw *flexWatcher) listOptions() *client.ListOptions {
+	opts := client.ListOptions{}
+	if len(fw.Key.Namespace) > 0 {
+		opts.Namespace = fw.Key.Namespace
+	}
+	if len(fw.Key.Name) > 0 {
+		opts.FieldSelector = fields.OneTermEqualSelector(metav1.ObjectNameField, fw.Key.Name)
+	}
+	if len(fw.Labels) > 0 {
+		opts.LabelSelector = client.MatchingLabelsSelector{
+			Selector: labels.SelectorFromSet(fw.Labels),
+		}
+	}
+	return &opts
+}
+
+// MergeListOptions merges two sets of ListOptions.
+// - For Namespace, Limit, and Continue, at most one may be specified, unless they're the same
+// - For FieldSelector and LabelSelector, the result requires both filters to match (AND)
+// - For Raw, at most one may be specified
+func MergeListOptions(a, b *client.ListOptions) (*client.ListOptions, error) {
+	switch {
+	case a == nil && b == nil:
+		return nil, nil
+	case a == nil:
+		return b, nil
+	case b == nil:
+		return a, nil
+	}
+
+	c := client.ListOptions{}
+
+	switch {
+	case a.Namespace != "" && b.Namespace != "":
+		if a.Namespace != b.Namespace {
+			return nil, errors.Errorf("cannot merge two different namespaces: %s & %s",
+				a.Namespace, b.Namespace)
+		}
+		c.Namespace = a.Namespace
+	case a.Namespace != "":
+		c.Namespace = a.Namespace
+	case b.Namespace != "":
+		c.Namespace = b.Namespace
+	}
+
+	switch {
+	case a.FieldSelector != nil && b.FieldSelector != nil:
+		c.FieldSelector = fields.AndSelectors(a.FieldSelector, b.FieldSelector)
+	case a.FieldSelector != nil:
+		c.FieldSelector = a.FieldSelector
+	case b.FieldSelector != nil:
+		c.FieldSelector = b.FieldSelector
+	}
+
+	switch {
+	case a.LabelSelector != nil && b.LabelSelector != nil:
+		bReqs, selectable := b.LabelSelector.Requirements()
+		if selectable {
+			c.LabelSelector = a.LabelSelector.Add(bReqs...)
+		} else {
```

Comment:
Is it desirable to silently drop `b.LabelSelector` in this case? Seems like that could hide issues, and we should return an error here - although I don't have full understanding of what `selectable` means

Reply:
`labels.internalSelector.Requirements()` basically returns self, because `internalSelector` is an alias for `[]Requirement`. And for some unknown reason it always returns true, even if the list is empty/nil. So this is probably always be selectable unless someone makes a different `labels.Requirements` impl.

Anyway, when it's selectable, the next line adds the "b" requirements to the "a" selector (which is itself a list of requirements).

Reply:
Looks like `nothingSelector.Requirements()` actually returns false. So maybe it doesn't mean what I thought it meant. If it's not selectable the merge would be the the nothingSelector.

Reply:
Fixed.